// nxkit/mysql/charsets.js
"use strict";
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */
Object.defineProperty(exports, "__esModule", { value: true });
var Charsets;
(function (Charsets) {
    Charsets[Charsets["BIG5_CHINESE_CI"] = 1] = "BIG5_CHINESE_CI";
    Charsets[Charsets["LATIN2_CZECH_CS"] = 2] = "LATIN2_CZECH_CS";
    Charsets[Charsets["DEC8_SWEDISH_CI"] = 3] = "DEC8_SWEDISH_CI";
    Charsets[Charsets["CP850_GENERAL_CI"] = 4] = "CP850_GENERAL_CI";
    Charsets[Charsets["LATIN1_GERMAN1_CI"] = 5] = "LATIN1_GERMAN1_CI";
    Charsets[Charsets["HP8_ENGLISH_CI"] = 6] = "HP8_ENGLISH_CI";
    Charsets[Charsets["KOI8R_GENERAL_CI"] = 7] = "KOI8R_GENERAL_CI";
    Charsets[Charsets["LATIN1_SWEDISH_CI"] = 8] = "LATIN1_SWEDISH_CI";
    Charsets[Charsets["LATIN2_GENERAL_CI"] = 9] = "LATIN2_GENERAL_CI";
    Charsets[Charsets["SWE7_SWEDISH_CI"] = 10] = "SWE7_SWEDISH_CI";
    Charsets[Charsets["ASCII_GENERAL_CI"] = 11] = "ASCII_GENERAL_CI";
    Charsets[Charsets["UJIS_JAPANESE_CI"] = 12] = "UJIS_JAPANESE_CI";
    Charsets[Charsets["SJIS_JAPANESE_CI"] = 13] = "SJIS_JAPANESE_CI";
    Charsets[Charsets["CP1251_BULGARIAN_CI"] = 14] = "CP1251_BULGARIAN_CI";
    Charsets[Charsets["LATIN1_DANISH_CI"] = 15] = "LATIN1_DANISH_CI";
    Charsets[Charsets["HEBREW_GENERAL_CI"] = 16] = "HEBREW_GENERAL_CI";
    Charsets[Charsets["TIS620_THAI_CI"] = 18] = "TIS620_THAI_CI";
    Charsets[Charsets["EUCKR_KOREAN_CI"] = 19] = "EUCKR_KOREAN_CI";
    Charsets[Charsets["LATIN7_ESTONIAN_CS"] = 20] = "LATIN7_ESTONIAN_CS";
    Charsets[Charsets["LATIN2_HUNGARIAN_CI"] = 21] = "LATIN2_HUNGARIAN_CI";
    Charsets[Charsets["KOI8U_GENERAL_CI"] = 22] = "KOI8U_GENERAL_CI";
    Charsets[Charsets["CP1251_UKRAINIAN_CI"] = 23] = "CP1251_UKRAINIAN_CI";
    Charsets[Charsets["GB2312_CHINESE_CI"] = 24] = "GB2312_CHINESE_CI";
    Charsets[Charsets["GREEK_GENERAL_CI"] = 25] = "GREEK_GENERAL_CI";
    Charsets[Charsets["CP1250_GENERAL_CI"] = 26] = "CP1250_GENERAL_CI";
    Charsets[Charsets["LATIN2_CROATIAN_CI"] = 27] = "LATIN2_CROATIAN_CI";
    Charsets[Charsets["GBK_CHINESE_CI"] = 28] = "GBK_CHINESE_CI";
    Charsets[Charsets["CP1257_LITHUANIAN_CI"] = 29] = "CP1257_LITHUANIAN_CI";
    Charsets[Charsets["LATIN5_TURKISH_CI"] = 30] = "LATIN5_TURKISH_CI";
    Charsets[Charsets["LATIN1_GERMAN2_CI"] = 31] = "LATIN1_GERMAN2_CI";
    Charsets[Charsets["ARMSCII8_GENERAL_CI"] = 32] = "ARMSCII8_GENERAL_CI";
    Charsets[Charsets["UTF8_GENERAL_CI"] = 33] = "UTF8_GENERAL_CI";
    Charsets[Charsets["CP1250_CZECH_CS"] = 34] = "CP1250_CZECH_CS";
    Charsets[Charsets["UCS2_GENERAL_CI"] = 35] = "UCS2_GENERAL_CI";
    Charsets[Charsets["CP866_GENERAL_CI"] = 36] = "CP866_GENERAL_CI";
    Charsets[Charsets["KEYBCS2_GENERAL_CI"] = 37] = "KEYBCS2_GENERAL_CI";
    Charsets[Charsets["MACCE_GENERAL_CI"] = 38] = "MACCE_GENERAL_CI";
    Charsets[Charsets["MACROMAN_GENERAL_CI"] = 39] = "MACROMAN_GENERAL_CI";
    Charsets[Charsets["CP852_GENERAL_CI"] = 40] = "CP852_GENERAL_CI";
    Charsets[Charsets["LATIN7_GENERAL_CI"] = 41] = "LATIN7_GENERAL_CI";
    Charsets[Charsets["LATIN7_GENERAL_CS"] = 42] = "LATIN7_GENERAL_CS";
    Charsets[Charsets["MACCE_BIN"] = 43] = "MACCE_BIN";
    Charsets[Charsets["CP1250_CROATIAN_CI"] = 44] = "CP1250_CROATIAN_CI";
    Charsets[Charsets["UTF8MB4_GENERAL_CI"] = 45] = "UTF8MB4_GENERAL_CI";
    Charsets[Charsets["UTF8MB4_BIN"] = 46] = "UTF8MB4_BIN";
    Charsets[Charsets["LATIN1_BIN"] = 47] = "LATIN1_BIN";
    Charsets[Charsets["LATIN1_GENERAL_CI"] = 48] = "LATIN1_GENERAL_CI";
    Charsets[Charsets["LATIN1_GENERAL_CS"] = 49] = "LATIN1_GENERAL_CS";
    Charsets[Charsets["CP1251_BIN"] = 50] = "CP1251_BIN";
    Charsets[Charsets["CP1251_GENERAL_CI"] = 51] = "CP1251_GENERAL_CI";
    Charsets[Charsets["CP1251_GENERAL_CS"] = 52] = "CP1251_GENERAL_CS";
    Charsets[Charsets["MACROMAN_BIN"] = 53] = "MACROMAN_BIN";
    Charsets[Charsets["UTF16_GENERAL_CI"] = 54] = "UTF16_GENERAL_CI";
    Charsets[Charsets["UTF16_BIN"] = 55] = "UTF16_BIN";
    Charsets[Charsets["UTF16LE_GENERAL_CI"] = 56] = "UTF16LE_GENERAL_CI";
    Charsets[Charsets["CP1256_GENERAL_CI"] = 57] = "CP1256_GENERAL_CI";
    Charsets[Charsets["CP1257_BIN"] = 58] = "CP1257_BIN";
    Charsets[Charsets["CP1257_GENERAL_CI"] = 59] = "CP1257_GENERAL_CI";
    Charsets[Charsets["UTF32_GENERAL_CI"] = 60] = "UTF32_GENERAL_CI";
    Charsets[Charsets["UTF32_BIN"] = 61] = "UTF32_BIN";
    Charsets[Charsets["UTF16LE_BIN"] = 62] = "UTF16LE_BIN";
    Charsets[Charsets["BINARY"] = 63] = "BINARY";
    Charsets[Charsets["ARMSCII8_BIN"] = 64] = "ARMSCII8_BIN";
    Charsets[Charsets["ASCII_BIN"] = 65] = "ASCII_BIN";
    Charsets[Charsets["CP1250_BIN"] = 66] = "CP1250_BIN";
    Charsets[Charsets["CP1256_BIN"] = 67] = "CP1256_BIN";
    Charsets[Charsets["CP866_BIN"] = 68] = "CP866_BIN";
    Charsets[Charsets["DEC8_BIN"] = 69] = "DEC8_BIN";
    Charsets[Charsets["GREEK_BIN"] = 70] = "GREEK_BIN";
    Charsets[Charsets["HEBREW_BIN"] = 71] = "HEBREW_BIN";
    Charsets[Charsets["HP8_BIN"] = 72] = "HP8_BIN";
    Charsets[Charsets["KEYBCS2_BIN"] = 73] = "KEYBCS2_BIN";
    Charsets[Charsets["KOI8R_BIN"] = 74] = "KOI8R_BIN";
    Charsets[Charsets["KOI8U_BIN"] = 75] = "KOI8U_BIN";
    Charsets[Charsets["LATIN2_BIN"] = 77] = "LATIN2_BIN";
    Charsets[Charsets["LATIN5_BIN"] = 78] = "LATIN5_BIN";
    Charsets[Charsets["LATIN7_BIN"] = 79] = "LATIN7_BIN";
    Charsets[Charsets["CP850_BIN"] = 80] = "CP850_BIN";
    Charsets[Charsets["CP852_BIN"] = 81] = "CP852_BIN";
    Charsets[Charsets["SWE7_BIN"] = 82] = "SWE7_BIN";
    Charsets[Charsets["UTF8_BIN"] = 83] = "UTF8_BIN";
    Charsets[Charsets["BIG5_BIN"] = 84] = "BIG5_BIN";
    Charsets[Charsets["EUCKR_BIN"] = 85] = "EUCKR_BIN";
    Charsets[Charsets["GB2312_BIN"] = 86] = "GB2312_BIN";
    Charsets[Charsets["GBK_BIN"] = 87] = "GBK_BIN";
    Charsets[Charsets["SJIS_BIN"] = 88] = "SJIS_BIN";
    Charsets[Charsets["TIS620_BIN"] = 89] = "TIS620_BIN";
    Charsets[Charsets["UCS2_BIN"] = 90] = "UCS2_BIN";
    Charsets[Charsets["UJIS_BIN"] = 91] = "UJIS_BIN";
    Charsets[Charsets["GEOSTD8_GENERAL_CI"] = 92] = "GEOSTD8_GENERAL_CI";
    Charsets[Charsets["GEOSTD8_BIN"] = 93] = "GEOSTD8_BIN";
    Charsets[Charsets["LATIN1_SPANISH_CI"] = 94] = "LATIN1_SPANISH_CI";
    Charsets[Charsets["CP932_JAPANESE_CI"] = 95] = "CP932_JAPANESE_CI";
    Charsets[Charsets["CP932_BIN"] = 96] = "CP932_BIN";
    Charsets[Charsets["EUCJPMS_JAPANESE_CI"] = 97] = "EUCJPMS_JAPANESE_CI";
    Charsets[Charsets["EUCJPMS_BIN"] = 98] = "EUCJPMS_BIN";
    Charsets[Charsets["CP1250_POLISH_CI"] = 99] = "CP1250_POLISH_CI";
    Charsets[Charsets["UTF16_UNICODE_CI"] = 101] = "UTF16_UNICODE_CI";
    Charsets[Charsets["UTF16_ICELANDIC_CI"] = 102] = "UTF16_ICELANDIC_CI";
    Charsets[Charsets["UTF16_LATVIAN_CI"] = 103] = "UTF16_LATVIAN_CI";
    Charsets[Charsets["UTF16_ROMANIAN_CI"] = 104] = "UTF16_ROMANIAN_CI";
    Charsets[Charsets["UTF16_SLOVENIAN_CI"] = 105] = "UTF16_SLOVENIAN_CI";
    Charsets[Charsets["UTF16_POLISH_CI"] = 106] = "UTF16_POLISH_CI";
    Charsets[Charsets["UTF16_ESTONIAN_CI"] = 107] = "UTF16_ESTONIAN_CI";
    Charsets[Charsets["UTF16_SPANISH_CI"] = 108] = "UTF16_SPANISH_CI";
    Charsets[Charsets["UTF16_SWEDISH_CI"] = 109] = "UTF16_SWEDISH_CI";
    Charsets[Charsets["UTF16_TURKISH_CI"] = 110] = "UTF16_TURKISH_CI";
    Charsets[Charsets["UTF16_CZECH_CI"] = 111] = "UTF16_CZECH_CI";
    Charsets[Charsets["UTF16_DANISH_CI"] = 112] = "UTF16_DANISH_CI";
    Charsets[Charsets["UTF16_LITHUANIAN_CI"] = 113] = "UTF16_LITHUANIAN_CI";
    Charsets[Charsets["UTF16_SLOVAK_CI"] = 114] = "UTF16_SLOVAK_CI";
    Charsets[Charsets["UTF16_SPANISH2_CI"] = 115] = "UTF16_SPANISH2_CI";
    Charsets[Charsets["UTF16_ROMAN_CI"] = 116] = "UTF16_ROMAN_CI";
    Charsets[Charsets["UTF16_PERSIAN_CI"] = 117] = "UTF16_PERSIAN_CI";
    Charsets[Charsets["UTF16_ESPERANTO_CI"] = 118] = "UTF16_ESPERANTO_CI";
    Charsets[Charsets["UTF16_HUNGARIAN_CI"] = 119] = "UTF16_HUNGARIAN_CI";
    Charsets[Charsets["UTF16_SINHALA_CI"] = 120] = "UTF16_SINHALA_CI";
    Charsets[Charsets["UTF16_GERMAN2_CI"] = 121] = "UTF16_GERMAN2_CI";
    Charsets[Charsets["UTF16_CROATIAN_MYSQL561_CI"] = 122] = "UTF16_CROATIAN_MYSQL561_CI";
    Charsets[Charsets["UTF16_UNICODE_520_CI"] = 123] = "UTF16_UNICODE_520_CI";
    Charsets[Charsets["UTF16_VIETNAMESE_CI"] = 124] = "UTF16_VIETNAMESE_CI";
    Charsets[Charsets["UCS2_UNICODE_CI"] = 128] = "UCS2_UNICODE_CI";
    Charsets[Charsets["UCS2_ICELANDIC_CI"] = 129] = "UCS2_ICELANDIC_CI";
    Charsets[Charsets["UCS2_LATVIAN_CI"] = 130] = "UCS2_LATVIAN_CI";
    Charsets[Charsets["UCS2_ROMANIAN_CI"] = 131] = "UCS2_ROMANIAN_CI";
    Charsets[Charsets["UCS2_SLOVENIAN_CI"] = 132] = "UCS2_SLOVENIAN_CI";
    Charsets[Charsets["UCS2_POLISH_CI"] = 133] = "UCS2_POLISH_CI";
    Charsets[Charsets["UCS2_ESTONIAN_CI"] = 134] = "UCS2_ESTONIAN_CI";
    Charsets[Charsets["UCS2_SPANISH_CI"] = 135] = "UCS2_SPANISH_CI";
    Charsets[Charsets["UCS2_SWEDISH_CI"] = 136] = "UCS2_SWEDISH_CI";
    Charsets[Charsets["UCS2_TURKISH_CI"] = 137] = "UCS2_TURKISH_CI";
    Charsets[Charsets["UCS2_CZECH_CI"] = 138] = "UCS2_CZECH_CI";
    Charsets[Charsets["UCS2_DANISH_CI"] = 139] = "UCS2_DANISH_CI";
    Charsets[Charsets["UCS2_LITHUANIAN_CI"] = 140] = "UCS2_LITHUANIAN_CI";
    Charsets[Charsets["UCS2_SLOVAK_CI"] = 141] = "UCS2_SLOVAK_CI";
    Charsets[Charsets["UCS2_SPANISH2_CI"] = 142] = "UCS2_SPANISH2_CI";
    Charsets[Charsets["UCS2_ROMAN_CI"] = 143] = "UCS2_ROMAN_CI";
    Charsets[Charsets["UCS2_PERSIAN_CI"] = 144] = "UCS2_PERSIAN_CI";
    Charsets[Charsets["UCS2_ESPERANTO_CI"] = 145] = "UCS2_ESPERANTO_CI";
    Charsets[Charsets["UCS2_HUNGARIAN_CI"] = 146] = "UCS2_HUNGARIAN_CI";
    Charsets[Charsets["UCS2_SINHALA_CI"] = 147] = "UCS2_SINHALA_CI";
    Charsets[Charsets["UCS2_GERMAN2_CI"] = 148] = "UCS2_GERMAN2_CI";
    Charsets[Charsets["UCS2_CROATIAN_MYSQL561_CI"] = 149] = "UCS2_CROATIAN_MYSQL561_CI";
    Charsets[Charsets["UCS2_UNICODE_520_CI"] = 150] = "UCS2_UNICODE_520_CI";
    Charsets[Charsets["UCS2_VIETNAMESE_CI"] = 151] = "UCS2_VIETNAMESE_CI";
    Charsets[Charsets["UCS2_GENERAL_MYSQL500_CI"] = 159] = "UCS2_GENERAL_MYSQL500_CI";
    Charsets[Charsets["UTF32_UNICODE_CI"] = 160] = "UTF32_UNICODE_CI";
    Charsets[Charsets["UTF32_ICELANDIC_CI"] = 161] = "UTF32_ICELANDIC_CI";
    Charsets[Charsets["UTF32_LATVIAN_CI"] = 162] = "UTF32_LATVIAN_CI";
    Charsets[Charsets["UTF32_ROMANIAN_CI"] = 163] = "UTF32_ROMANIAN_CI";
    Charsets[Charsets["UTF32_SLOVENIAN_CI"] = 164] = "UTF32_SLOVENIAN_CI";
    Charsets[Charsets["UTF32_POLISH_CI"] = 165] = "UTF32_POLISH_CI";
    Charsets[Charsets["UTF32_ESTONIAN_CI"] = 166] = "UTF32_ESTONIAN_CI";
    Charsets[Charsets["UTF32_SPANISH_CI"] = 167] = "UTF32_SPANISH_CI";
    Charsets[Charsets["UTF32_SWEDISH_CI"] = 168] = "UTF32_SWEDISH_CI";
    Charsets[Charsets["UTF32_TURKISH_CI"] = 169] = "UTF32_TURKISH_CI";
    Charsets[Charsets["UTF32_CZECH_CI"] = 170] = "UTF32_CZECH_CI";
    Charsets[Charsets["UTF32_DANISH_CI"] = 171] = "UTF32_DANISH_CI";
    Charsets[Charsets["UTF32_LITHUANIAN_CI"] = 172] = "UTF32_LITHUANIAN_CI";
    Charsets[Charsets["UTF32_SLOVAK_CI"] = 173] = "UTF32_SLOVAK_CI";
    Charsets[Charsets["UTF32_SPANISH2_CI"] = 174] = "UTF32_SPANISH2_CI";
    Charsets[Charsets["UTF32_ROMAN_CI"] = 175] = "UTF32_ROMAN_CI";
    Charsets[Charsets["UTF32_PERSIAN_CI"] = 176] = "UTF32_PERSIAN_CI";
    Charsets[Charsets["UTF32_ESPERANTO_CI"] = 177] = "UTF32_ESPERANTO_CI";
    Charsets[Charsets["UTF32_HUNGARIAN_CI"] = 178] = "UTF32_HUNGARIAN_CI";
    Charsets[Charsets["UTF32_SINHALA_CI"] = 179] = "UTF32_SINHALA_CI";
    Charsets[Charsets["UTF32_GERMAN2_CI"] = 180] = "UTF32_GERMAN2_CI";
    Charsets[Charsets["UTF32_CROATIAN_MYSQL561_CI"] = 181] = "UTF32_CROATIAN_MYSQL561_CI";
    Charsets[Charsets["UTF32_UNICODE_520_CI"] = 182] = "UTF32_UNICODE_520_CI";
    Charsets[Charsets["UTF32_VIETNAMESE_CI"] = 183] = "UTF32_VIETNAMESE_CI";
    Charsets[Charsets["UTF8_UNICODE_CI"] = 192] = "UTF8_UNICODE_CI";
    Charsets[Charsets["UTF8_ICELANDIC_CI"] = 193] = "UTF8_ICELANDIC_CI";
    Charsets[Charsets["UTF8_LATVIAN_CI"] = 194] = "UTF8_LATVIAN_CI";
    Charsets[Charsets["UTF8_ROMANIAN_CI"] = 195] = "UTF8_ROMANIAN_CI";
    Charsets[Charsets["UTF8_SLOVENIAN_CI"] = 196] = "UTF8_SLOVENIAN_CI";
    Charsets[Charsets["UTF8_POLISH_CI"] = 197] = "UTF8_POLISH_CI";
    Charsets[Charsets["UTF8_ESTONIAN_CI"] = 198] = "UTF8_ESTONIAN_CI";
    Charsets[Charsets["UTF8_SPANISH_CI"] = 199] = "UTF8_SPANISH_CI";
    Charsets[Charsets["UTF8_SWEDISH_CI"] = 200] = "UTF8_SWEDISH_CI";
    Charsets[Charsets["UTF8_TURKISH_CI"] = 201] = "UTF8_TURKISH_CI";
    Charsets[Charsets["UTF8_CZECH_CI"] = 202] = "UTF8_CZECH_CI";
    Charsets[Charsets["UTF8_DANISH_CI"] = 203] = "UTF8_DANISH_CI";
    Charsets[Charsets["UTF8_LITHUANIAN_CI"] = 204] = "UTF8_LITHUANIAN_CI";
    Charsets[Charsets["UTF8_SLOVAK_CI"] = 205] = "UTF8_SLOVAK_CI";
    Charsets[Charsets["UTF8_SPANISH2_CI"] = 206] = "UTF8_SPANISH2_CI";
    Charsets[Charsets["UTF8_ROMAN_CI"] = 207] = "UTF8_ROMAN_CI";
    Charsets[Charsets["UTF8_PERSIAN_CI"] = 208] = "UTF8_PERSIAN_CI";
    Charsets[Charsets["UTF8_ESPERANTO_CI"] = 209] = "UTF8_ESPERANTO_CI";
    Charsets[Charsets["UTF8_HUNGARIAN_CI"] = 210] = "UTF8_HUNGARIAN_CI";
    Charsets[Charsets["UTF8_SINHALA_CI"] = 211] = "UTF8_SINHALA_CI";
    Charsets[Charsets["UTF8_GERMAN2_CI"] = 212] = "UTF8_GERMAN2_CI";
    Charsets[Charsets["UTF8_CROATIAN_MYSQL561_CI"] = 213] = "UTF8_CROATIAN_MYSQL561_CI";
    Charsets[Charsets["UTF8_UNICODE_520_CI"] = 214] = "UTF8_UNICODE_520_CI";
    Charsets[Charsets["UTF8_VIETNAMESE_CI"] = 215] = "UTF8_VIETNAMESE_CI";
    Charsets[Charsets["UTF8_GENERAL_MYSQL500_CI"] = 223] = "UTF8_GENERAL_MYSQL500_CI";
    Charsets[Charsets["UTF8MB4_UNICODE_CI"] = 224] = "UTF8MB4_UNICODE_CI";
    Charsets[Charsets["UTF8MB4_ICELANDIC_CI"] = 225] = "UTF8MB4_ICELANDIC_CI";
    Charsets[Charsets["UTF8MB4_LATVIAN_CI"] = 226] = "UTF8MB4_LATVIAN_CI";
    Charsets[Charsets["UTF8MB4_ROMANIAN_CI"] = 227] = "UTF8MB4_ROMANIAN_CI";
    Charsets[Charsets["UTF8MB4_SLOVENIAN_CI"] = 228] = "UTF8MB4_SLOVENIAN_CI";
    Charsets[Charsets["UTF8MB4_POLISH_CI"] = 229] = "UTF8MB4_POLISH_CI";
    Charsets[Charsets["UTF8MB4_ESTONIAN_CI"] = 230] = "UTF8MB4_ESTONIAN_CI";
    Charsets[Charsets["UTF8MB4_SPANISH_CI"] = 231] = "UTF8MB4_SPANISH_CI";
    Charsets[Charsets["UTF8MB4_SWEDISH_CI"] = 232] = "UTF8MB4_SWEDISH_CI";
    Charsets[Charsets["UTF8MB4_TURKISH_CI"] = 233] = "UTF8MB4_TURKISH_CI";
    Charsets[Charsets["UTF8MB4_CZECH_CI"] = 234] = "UTF8MB4_CZECH_CI";
    Charsets[Charsets["UTF8MB4_DANISH_CI"] = 235] = "UTF8MB4_DANISH_CI";
    Charsets[Charsets["UTF8MB4_LITHUANIAN_CI"] = 236] = "UTF8MB4_LITHUANIAN_CI";
    Charsets[Charsets["UTF8MB4_SLOVAK_CI"] = 237] = "UTF8MB4_SLOVAK_CI";
    Charsets[Charsets["UTF8MB4_SPANISH2_CI"] = 238] = "UTF8MB4_SPANISH2_CI";
    Charsets[Charsets["UTF8MB4_ROMAN_CI"] = 239] = "UTF8MB4_ROMAN_CI";
    Charsets[Charsets["UTF8MB4_PERSIAN_CI"] = 240] = "UTF8MB4_PERSIAN_CI";
    Charsets[Charsets["UTF8MB4_ESPERANTO_CI"] = 241] = "UTF8MB4_ESPERANTO_CI";
    Charsets[Charsets["UTF8MB4_HUNGARIAN_CI"] = 242] = "UTF8MB4_HUNGARIAN_CI";
    Charsets[Charsets["UTF8MB4_SINHALA_CI"] = 243] = "UTF8MB4_SINHALA_CI";
    Charsets[Charsets["UTF8MB4_GERMAN2_CI"] = 244] = "UTF8MB4_GERMAN2_CI";
    Charsets[Charsets["UTF8MB4_CROATIAN_MYSQL561_CI"] = 245] = "UTF8MB4_CROATIAN_MYSQL561_CI";
    Charsets[Charsets["UTF8MB4_UNICODE_520_CI"] = 246] = "UTF8MB4_UNICODE_520_CI";
    Charsets[Charsets["UTF8MB4_VIETNAMESE_CI"] = 247] = "UTF8MB4_VIETNAMESE_CI";
    Charsets[Charsets["UTF8_GENERAL50_CI"] = 253] = "UTF8_GENERAL50_CI";
    // short aliases
    Charsets[Charsets["ARMSCII8"] = 32] = "ARMSCII8";
    Charsets[Charsets["ASCII"] = 11] = "ASCII";
    Charsets[Charsets["BIG5"] = 1] = "BIG5";
    Charsets[Charsets["CP1250"] = 26] = "CP1250";
    Charsets[Charsets["CP1251"] = 51] = "CP1251";
    Charsets[Charsets["CP1256"] = 57] = "CP1256";
    Charsets[Charsets["CP1257"] = 59] = "CP1257";
    Charsets[Charsets["CP866"] = 36] = "CP866";
    Charsets[Charsets["CP850"] = 4] = "CP850";
    Charsets[Charsets["CP852"] = 40] = "CP852";
    Charsets[Charsets["CP932"] = 95] = "CP932";
    Charsets[Charsets["DEC8"] = 3] = "DEC8";
    Charsets[Charsets["EUCJPMS"] = 97] = "EUCJPMS";
    Charsets[Charsets["EUCKR"] = 19] = "EUCKR";
    Charsets[Charsets["GB2312"] = 24] = "GB2312";
    Charsets[Charsets["GBK"] = 28] = "GBK";
    Charsets[Charsets["GEOSTD8"] = 92] = "GEOSTD8";
    Charsets[Charsets["GREEK"] = 25] = "GREEK";
    Charsets[Charsets["HEBREW"] = 16] = "HEBREW";
    Charsets[Charsets["HP8"] = 6] = "HP8";
    Charsets[Charsets["KEYBCS2"] = 37] = "KEYBCS2";
    Charsets[Charsets["KOI8R"] = 7] = "KOI8R";
    Charsets[Charsets["KOI8U"] = 22] = "KOI8U";
    Charsets[Charsets["LATIN1"] = 8] = "LATIN1";
    Charsets[Charsets["LATIN2"] = 9] = "LATIN2";
    Charsets[Charsets["LATIN5"] = 30] = "LATIN5";
    Charsets[Charsets["LATIN7"] = 41] = "LATIN7";
    Charsets[Charsets["MACCE"] = 38] = "MACCE";
    Charsets[Charsets["MACROMAN"] = 39] = "MACROMAN";
    Charsets[Charsets["SJIS"] = 13] = "SJIS";
    Charsets[Charsets["SWE7"] = 10] = "SWE7";
    Charsets[Charsets["TIS620"] = 18] = "TIS620";
    Charsets[Charsets["UCS2"] = 35] = "UCS2";
    Charsets[Charsets["UJIS"] = 12] = "UJIS";
    Charsets[Charsets["UTF16"] = 54] = "UTF16";
    Charsets[Charsets["UTF16LE"] = 56] = "UTF16LE";
    Charsets[Charsets["UTF8"] = 33] = "UTF8";
    Charsets[Charsets["UTF8MB4"] = 45] = "UTF8MB4";
    Charsets[Charsets["UTF32"] = 60] = "UTF32";
})(Charsets = exports.Charsets || (exports.Charsets = {}));
;
exports.default = Charsets;
