// nxkit/mysql/constants.js
"use strict";
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */
Object.defineProperty(exports, "__esModule", { value: true });
var Constants;
(function (Constants) {
    // Connections Flags
    // Manually extracted from mysql-5.5.23/include/mysql_com.h
    Constants[Constants["CLIENT_LONG_PASSWORD"] = 1] = "CLIENT_LONG_PASSWORD";
    Constants[Constants["CLIENT_FOUND_ROWS"] = 2] = "CLIENT_FOUND_ROWS";
    Constants[Constants["CLIENT_LONG_FLAG"] = 4] = "CLIENT_LONG_FLAG";
    Constants[Constants["CLIENT_CONNECT_WITH_DB"] = 8] = "CLIENT_CONNECT_WITH_DB";
    Constants[Constants["CLIENT_NO_SCHEMA"] = 16] = "CLIENT_NO_SCHEMA";
    Constants[Constants["CLIENT_COMPRESS"] = 32] = "CLIENT_COMPRESS";
    Constants[Constants["CLIENT_ODBC"] = 64] = "CLIENT_ODBC";
    Constants[Constants["CLIENT_LOCAL_FILES"] = 128] = "CLIENT_LOCAL_FILES";
    Constants[Constants["CLIENT_IGNORE_SPACE"] = 256] = "CLIENT_IGNORE_SPACE";
    Constants[Constants["CLIENT_PROTOCOL_41"] = 512] = "CLIENT_PROTOCOL_41";
    Constants[Constants["CLIENT_INTERACTIVE"] = 1024] = "CLIENT_INTERACTIVE";
    Constants[Constants["CLIENT_SSL"] = 2048] = "CLIENT_SSL";
    Constants[Constants["CLIENT_IGNORE_SIGPIPE"] = 4096] = "CLIENT_IGNORE_SIGPIPE";
    Constants[Constants["CLIENT_TRANSACTIONS"] = 8192] = "CLIENT_TRANSACTIONS";
    Constants[Constants["CLIENT_RESERVED"] = 16384] = "CLIENT_RESERVED";
    Constants[Constants["CLIENT_SECURE_CONNECTION"] = 32768] = "CLIENT_SECURE_CONNECTION";
    Constants[Constants["CLIENT_MULTI_STATEMENTS"] = 65536] = "CLIENT_MULTI_STATEMENTS";
    Constants[Constants["CLIENT_MULTI_RESULTS"] = 131072] = "CLIENT_MULTI_RESULTS";
    Constants[Constants["CLIENT_PS_MULTI_RESULTS"] = 262144] = "CLIENT_PS_MULTI_RESULTS";
    Constants[Constants["CLIENT_PLUGIN_AUTH"] = 524288] = "CLIENT_PLUGIN_AUTH";
    Constants[Constants["CLIENT_SSL_VERIFY_SERVER_CERT"] = 1073741824] = "CLIENT_SSL_VERIFY_SERVER_CERT";
    Constants[Constants["CLIENT_REMEMBER_OPTIONS"] = 2147483648] = "CLIENT_REMEMBER_OPTIONS";
    // Commands
    Constants[Constants["COM_SLEEP"] = 0] = "COM_SLEEP";
    Constants[Constants["COM_QUIT"] = 1] = "COM_QUIT";
    Constants[Constants["COM_INIT_DB"] = 2] = "COM_INIT_DB";
    Constants[Constants["COM_QUERY"] = 3] = "COM_QUERY";
    Constants[Constants["COM_FIELD_LIST"] = 4] = "COM_FIELD_LIST";
    Constants[Constants["COM_CREATE_DB"] = 5] = "COM_CREATE_DB";
    Constants[Constants["COM_DROP_DB"] = 6] = "COM_DROP_DB";
    Constants[Constants["COM_REFRESH"] = 7] = "COM_REFRESH";
    Constants[Constants["COM_SHUTDOWN"] = 8] = "COM_SHUTDOWN";
    Constants[Constants["COM_STATISTICS"] = 9] = "COM_STATISTICS";
    Constants[Constants["COM_PROCESS_INFO"] = 10] = "COM_PROCESS_INFO";
    Constants[Constants["COM_CONNECT"] = 11] = "COM_CONNECT";
    Constants[Constants["COM_PROCESS_KILL"] = 12] = "COM_PROCESS_KILL";
    Constants[Constants["COM_DEBUG"] = 13] = "COM_DEBUG";
    Constants[Constants["COM_PING"] = 14] = "COM_PING";
    Constants[Constants["COM_TIME"] = 15] = "COM_TIME";
    Constants[Constants["COM_DELAYED_INSERT"] = 16] = "COM_DELAYED_INSERT";
    Constants[Constants["COM_CHANGE_USER"] = 17] = "COM_CHANGE_USER";
    Constants[Constants["COM_BINLOG_DUMP"] = 18] = "COM_BINLOG_DUMP";
    Constants[Constants["COM_TABLE_DUMP"] = 19] = "COM_TABLE_DUMP";
    Constants[Constants["COM_CONNECT_OUT"] = 20] = "COM_CONNECT_OUT";
    Constants[Constants["COM_REGISTER_SLAVE"] = 21] = "COM_REGISTER_SLAVE";
    Constants[Constants["COM_STMT_PREPARE"] = 22] = "COM_STMT_PREPARE";
    Constants[Constants["COM_STMT_EXECUTE"] = 23] = "COM_STMT_EXECUTE";
    Constants[Constants["COM_STMT_SEND_LONG_DATA"] = 24] = "COM_STMT_SEND_LONG_DATA";
    Constants[Constants["COM_STMT_CLOSE"] = 25] = "COM_STMT_CLOSE";
    Constants[Constants["COM_STMT_RESET"] = 26] = "COM_STMT_RESET";
    Constants[Constants["COM_SET_OPTION"] = 27] = "COM_SET_OPTION";
    Constants[Constants["COM_STMT_FETCH"] = 28] = "COM_STMT_FETCH";
})(Constants = exports.Constants || (exports.Constants = {}));
exports.default = Constants;
